internal/mcoa/mcoa.go

```diff
@@ -0,0 +1,104 @@
+package mcoa
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/rhobs/multicluster-observability-addon/internal/addon"
+	addonhelm "github.com/rhobs/multicluster-observability-addon/internal/addon/helm"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+	"k8s.io/kubectl/pkg/scheme"
+	"open-cluster-management.io/addon-framework/pkg/addonfactory"
+	"open-cluster-management.io/addon-framework/pkg/addonmanager"
+	"open-cluster-management.io/addon-framework/pkg/agent"
+	"open-cluster-management.io/addon-framework/pkg/utils"
+	addonv1alpha1client "open-cluster-management.io/api/client/addon/clientset/versioned"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+)
+
+type MCOA struct {
+	kubeConfig         *rest.Config
+	registrationOption *agent.RegistrationOption
+	mgr                *addonmanager.AddonManager
+	ctx                *context.Context
+}
+
+func NewMCOA(kubeConfig *rest.Config, registrationOption *agent.RegistrationOption) (*MCOA, error) {
+	mcoa := &MCOA{
+		kubeConfig:         kubeConfig,
+		registrationOption: registrationOption,
+	}
+	err := mcoa.Init()
+	if err != nil {
+		return nil, err
+	}
+
+	return mcoa, nil
+}
+
+func (m *MCOA) Init() error {
+	addonClient, err := addonv1alpha1client.NewForConfig(m.kubeConfig)
```

Comment:
Very likely, it would be a good idea to protect this from being run from different go routines.